google/cloud/aiplatform/pipeline_service/pipeline_based_service.py

```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import abc
+from datetime import datetime
+
+from google.auth import credentials as auth_credentials
+from google.protobuf import field_mask_pb2
+
+from google.cloud.aiplatform import base
+from google.cloud.aiplatform import initializer
+from google.cloud.aiplatform import utils
+from google.cloud.aiplatform import pipeline_jobs
+
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Iterable,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    Union,
+)
+
+from google.cloud.aiplatform.compat.types import (
+    pipeline_job_v1 as gca_pipeline_job_v1,
+    pipeline_state_v1 as gca_pipeline_state_v1,
+)
+
+_LOGGER = base.Logger(__name__)
+
+_PIPELINE_COMPLETE_STATES = set(
+    [
+        gca_pipeline_state_v1.PipelineState.PIPELINE_STATE_SUCCEEDED,
+        gca_pipeline_state_v1.PipelineState.PIPELINE_STATE_FAILED,
+        gca_pipeline_state_v1.PipelineState.PIPELINE_STATE_CANCELLED,
+        gca_pipeline_state_v1.PipelineState.PIPELINE_STATE_PAUSED,
+    ]
+)
+
+
+class VertexAiPipelineBasedService(base.VertexAiStatefulResource):
+    """Base class for Vertex AI Pipeline based services."""
+
+    client_class = utils.PipelineJobClientWithOverride
+    _resource_noun = "pipelineJob"
+    _delete_method = "delete_pipeline_job"
+    _getter_method = "get_pipeline_job"
+    _list_method = "list_pipeline_jobs"
+    _parse_resource_name_method = "parse_pipeline_job_path"
+    _format_resource_name_method = "pipeline_job_path"
+
+    _valid_done_states = _PIPELINE_COMPLETE_STATES
+
+    @property
+    @abc.abstractmethod
+    def backing_pipeline_job(self) -> pipeline_jobs.PipelineJob:
+        """The PipelineJob associated with the resource."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def pipeline_console_uri(self) -> str:
+        """The console URI of the PipelineJob created by the service."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def metadata_output_artifact(self) -> Optional[str]:
+        """The ML Metadata output artifact resource URI from the completed pipeline run."""
+        pass
+
+    @property
+    def state(self) -> Optional[str]:
+        """The state of the Pipeline run associated with the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job.state
```

Comment:
This is referencing `backing_pipeline_job` while it's currently being stored as `_gca_resource`